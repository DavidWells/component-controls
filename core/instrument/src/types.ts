import { ParserOptions } from '@babel/parser';
import { SyncOpts as ResolveOptions } from 'resolve';
import {
  Options,
  ResolveConfigOptions as ResolvePrettierConfigOptions,
} from 'prettier';
import { PropTypes } from '@component-controls/specification';

type PrettierOptions = Options & {
  resolveConfigOptions?: ResolvePrettierConfigOptions;
};
export { ParserOptions, ResolveOptions, PrettierOptions };

export const defaultResolveOptions: ResolveOptions = {
  extensions: ['.js', '.jsx', '.ts', '.tsx', '.vue', '.mjs', '.es', '.es6'],
};

export const defaultParserOptions: ParserOptions = {
  sourceType: 'module',
  plugins: ['jsx', 'typescript'],
};

/**
 * callback function to extract props info table  - ie docgen type libraries
 * used to extract displayName, and props tables for a component
 */
export type PropsInfoExtractor = (
  /**
   * full name and path of the component path
   * react-docgen needs it to extract babel configurations
   */
  fileName: string,
  /**
   * optional component name
   * react-docgen-typescript supports multiple exports for a file
   * react-docgne does not use it
   */
  componentName?: string,
  /**
   * optional soure, saves time if its already loaded
   * react-docgen accepts source as input parameter
   * react-docgen-typescript does not use it
   */
  source?: string,
) => PropTypes | undefined;

export const defaultPackageOptions: PackageInfoOptions = {
  maxLevels: 10,
  packageJsonName: 'package.json',
  storeBrowseLink: true,
  storeDocsLink: true,
  storeIssuesLink: true,
};

export const defaultComponentOptions: ComponentOptions = {
  storeSourceFile: true,
  package: defaultPackageOptions,
};

export const defaultStoriesOptions: StoriesOptions = {
  storeSourceFile: true,
  package: defaultPackageOptions,
};

/**
 * options for finding and extracting package.json informtation
 * applies both for components and stories
 */
export interface PackageInfoOptions {
  /**
   * max levels of folders to look up to find the package.json file
   */
  maxLevels?: number;
  /**
   * package.json alternative name
   */
  packageJsonName?: string;

  /**
   * Whether to save the link for browsing the file in the repository field
   */
  storeBrowseLink?: boolean;

  /**
   * Whether to save the link for project readme file in the repository field
   */
  storeDocsLink?: boolean;

  /**
   * Whether to save the link for filing issues with the project in the repository field
   */

  storeIssuesLink?: boolean;
}

export interface ComponentOptions {
  /**
   * Callback function to resolve the source file name of a component.
   * Return false if this file should not be processed.
   */
  resolveFile?: (componentName: string, filePath: string) => string | undefined;

  /**
   * optional module to extract prop tables information for components
   */
  extractProps?: PropsInfoExtractor;
  /**
   * If set to false, will not save the component's source file
   */
  storeSourceFile?: boolean;

  /**
   * options for extracting repository information from the component's package,json file
   */
  package?: PackageInfoOptions | false;
}

export interface StoriesOptions {
  /**
   * If set to false, will not save the stories's source file, only the source of each individual story
   */
  storeSourceFile?: boolean;

  /**
   * options for extracting repository information from the component's package,json file
   */
  package?: PackageInfoOptions | false;
}

/**
 * Options passed for instrumenting
 */
export interface InstrumentOptions {
  /**
   * Options to control babel parser.
   */
  parser?: ParserOptions;
  /**
   * prettier options are used to prettify the code at the end of the process
   * this is useful for "story" code, where the story is extracted from the full file
   * fassing a value of false as prettier option will disabled prettifying
   */
  prettier?: PrettierOptions | false;

  /**
   * Options to control the `resolve` import resolving utilities.
   */
  resolver?: ResolveOptions;

  /**
   * Options for extracting components information.
   */
  components?: ComponentOptions;

  /**
   * Options for extracting stories information.
   */
  stories?: StoriesOptions;
}

/**
 * mdx-js options
 */
export interface MDXOptions {
  /**
   * enable footnotes
   */

  footnotes?: boolean;
  /**
   * specify remark plugins
   */
  mdPlugins?: any[];
  /**
   * specify rehype plugins
   */
  hastPlugins?: any[];
  /**
   * specify markdown compilers
   */
  compilers?: any[];

  /**
   * regex for blocks, defaults to ['[a-z\\.]+(\\.){0,1}[a-z\\.]']
   */
  blocks?: string[];
}

/**
 * instrument MDX has additional mdx options
 */
export interface InstrumentOptionsMDX extends InstrumentOptions {
  mdx?: MDXOptions;
}
