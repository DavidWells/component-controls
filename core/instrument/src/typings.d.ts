declare module 'read-package-json';
declare module '@hutson/parse-repository-url';
declare module 'hosted-git-info';
declare module '@mdx-js/mdx';
declare module 'js-string-escape';
declare module 'remark-emoji';
declare module 'remark-images';
