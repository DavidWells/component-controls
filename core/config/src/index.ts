import * as path from 'path';
import * as fs from 'fs';
import { sync as globSync } from 'glob';
import yargs from 'yargs';
import { Configuration } from '@component-controls/specification';

export const configFileNames = ['main.js', 'main.json', 'main.ts'];

export interface ConfigrationResult {
  config: Configuration;
  configPath: string;
}
/**
 * return the configration object
 * command line accepts -c/ -config parameter for config path
 * the config file is assumed named main.js/main.ts
 * @param baseFolder project folder to start the searh with
 */
export const getConfiguration = (
  baseFolder: string,
  args: string[] = process.argv,
): ConfigrationResult | undefined => {
  const argv = yargs(args)
    .option('config', {
      alias: 'c',
      description: 'configuration folder',
      type: 'string',
    })
    .help()
    .alias('help', 'h').argv;

  const configPath = argv.config
    ? path.resolve(baseFolder, argv.config)
    : baseFolder;
  const configFiles = fs.readdirSync(configPath);
  const configFile = configFiles.find(file =>
    configFileNames.includes(file.toLowerCase()),
  );

  return configFile
    ? {
        config: require(path.resolve(configPath, configFile)),
        configPath,
      }
    : undefined;
};

/**
 * find the story files out of a configuration file
 * using glob for the regex file search
 * @param param0 configuration info return by getConfiguration
 */
export const extractStories = ({
  config,
  configPath,
}: ConfigrationResult): string[] | undefined => {
  const stories =
    config && config.stories
      ? config.stories.reduce((acc: string[], storyRg: string) => {
          const regex = storyRg.replace(/\((.*?)\)/g, '+($1)');
          return [...acc, ...globSync(path.resolve(configPath, regex))];
        }, [])
      : undefined;
  return stories;
};
