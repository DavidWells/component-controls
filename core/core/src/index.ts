export * from './utils';
export { randomizeData } from './randomizeData';
export * from './smartControls';
export * from './source';
export * from './faker';
