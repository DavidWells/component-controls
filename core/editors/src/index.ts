export * from './editors/prop-factory';
export * from './editors/types';
export * from './forms/ControlsTable/ControlEditorsTable';
export { FlexContainer } from './components/FlexContainer/FlexContainer';
