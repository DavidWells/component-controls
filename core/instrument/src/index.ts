import * as parser from '@babel/parser';
const mdx = require('@mdx-js/mdx');
import traverse from '@babel/traverse';
import generate from '@babel/generator';
import prettier, { Options, ResolveConfigOptions } from 'prettier';
import parserBabel from 'prettier/parser-babylon';
import { StoriesGroup, Story } from '@component-controls/specification';
import { extractCSFStories } from './babel/csf-stories';
import { extractMDXStories } from './babel/mdx-stories';
import { removeMDXAttributes } from './babel/remove-mdx-attributes';

type TraverseFn = (stories: StoriesGroup) => any;

export type PrettierOptions = Options & {
  resolveConfigOptions?: ResolveConfigOptions;
};
const parseSource = async (
  code: string,
  traverseFn: TraverseFn,
  originalSource: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  let source: string;
  if (prettierOptions !== false) {
    const { resolveConfigOptions, ...otherOptions } = prettierOptions || {};
    let allPrettierOptions = otherOptions;
    if (filePath) {
      const userOptions = await prettier.resolveConfig(
        filePath,
        resolveConfigOptions,
      );
      allPrettierOptions = { ...userOptions, ...allPrettierOptions };
    }

    source = prettier.format(code, {
      parser: 'typescript',
      plugins: [parserBabel],
      ...allPrettierOptions,
    });
  } else {
    source = code;
  }

  const ast = parser.parse(source, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript'],
  });
  const stories = {
    stories: {},
    source: originalSource,
  };
  traverse(ast, traverseFn(stories));
  Object.keys(stories.stories).forEach((key: string) => {
    //@ts-ignore
    const story: Story = stories.stories[key];
    const { start, end } = story.loc || {};
    if (start && end) {
      const lines = source.split('\n');

      if (start.line === end.line) {
        story.source = lines[start.line - 1].substring(
          start.column,
          end.column,
        );
      } else {
        const startLine = lines[start.line - 1];
        const endLine = lines[end.line - 1];
        if (startLine !== undefined && endLine !== undefined) {
          story.source = [
            startLine.substring(start.column),
            ...lines.slice(start.line, end.line - 1),
            endLine.substring(0, end.column),
          ].join('\n');
        }
      }
    }
  });
  return stories;
};

export const parseCSF = async (
  source: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  return await parseSource(
    source,
    extractCSFStories,
    source,
    filePath,
    prettierOptions,
  );
};

export const parseMDX = async (
  source: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  const code = await mdx(source);

  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript'],
  });
  traverse(ast, removeMDXAttributes());
  const newCode = generate(ast, {
    retainFunctionParens: true,
    retainLines: true,
  });
  return await parseSource(
    newCode.code,
    extractMDXStories,
    source,
    filePath,
    prettierOptions,
  );
};
