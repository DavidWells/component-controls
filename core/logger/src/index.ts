import chalk from 'chalk';

export type LogLevel = 'all' | 'errors' | 'none';

export interface LogOptions {
  logLevel?: LogLevel;
  colors: [number, number, number];
}

const defaultOptions: LogOptions = {
  logLevel: 'all',
  colors: [244, 147, 66],
};

let logOptions: LogOptions = { ...defaultOptions };

const output = (heading: string, text: string) => {
  console.log(chalk.bgRgb(...logOptions.colors)(`@${heading}`), text);
};

export const log = (heading: string, text: string): void => {
  if (logOptions.logLevel === 'all') {
    output(heading, text);
  }
};
export const error = (heading: string, text: string): void => {
  if (logOptions.logLevel !== 'none') {
    output(heading, text);
  }
};

export const setLogOptions = (options: Partial<LogOptions> = {}): void => {
  logOptions = { ...logOptions, ...options };
};
