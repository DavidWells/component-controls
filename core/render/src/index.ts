import { FrameworkRenderFn } from '@component-controls/core';
import { render as renderReact } from './react';

//noop
export const renderers: Record<string, FrameworkRenderFn> = {
  react: renderReact,
};
