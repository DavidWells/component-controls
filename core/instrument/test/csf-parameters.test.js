import { parseCSF } from '../src/index';

describe('csf-parameters', () => {
  it('story propery name', () => {
    expect(
      parseCSF(`
export const myStory = () => {};
myStory.story = {
 name: 'Custom story name',
}
      
    `),
    ).toMatchSnapshot();
  });

  it('story propery name and parameters', () => {
    expect(
      parseCSF(`
export const myStory = () => {};
myStory.story = {
  name: 'Custom story name',
  parameters: {
    component: ControlsEditorsTable,
    addonControls: {
      smart: false,
    }
  },
}
      
    `),
    ).toMatchSnapshot();
  });

  it('story propery name parameters and controls', () => {
    expect(
      parseCSF(`
export const myStory = () => {};
myStory.story = {
  name: 'Custom story name',
  parameters: {
    controls: {
      name: {
        type: ControlTypes.TEXT,
        label: 'Name',
        value: 'Mark',
        order: 9999,
      },
    },
    addonControls: {
      smart: false,
    }
  },
}
      
    `),
    ).toMatchSnapshot();
  });
});
