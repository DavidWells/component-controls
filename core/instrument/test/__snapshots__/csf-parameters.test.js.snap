// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csf-parameters story propery name 1`] = `
Object {
  "stories": Object {
    "myStory": Object {
      "arguments": Array [],
      "location": Object {
        "end": Object {
          "column": 31,
          "line": 2,
        },
        "start": Object {
          "column": 23,
          "line": 2,
        },
      },
      "name": "Custom story name",
      "properties": Array [
        Object {
          "loc": SourceLocation {
            "end": Position {
              "column": 26,
              "line": 4,
            },
            "start": Position {
              "column": 1,
              "line": 4,
            },
          },
          "name": "name",
          "value": "Custom story name",
        },
      ],
      "source": "() => {}",
    },
  },
}
`;

exports[`csf-parameters story propery name and parameters 1`] = `
Object {
  "stories": Object {
    "myStory": Object {
      "arguments": Array [],
      "location": Object {
        "end": Object {
          "column": 31,
          "line": 2,
        },
        "start": Object {
          "column": 23,
          "line": 2,
        },
      },
      "name": "Custom story name",
      "properties": Array [
        Object {
          "loc": SourceLocation {
            "end": Position {
              "column": 27,
              "line": 4,
            },
            "start": Position {
              "column": 2,
              "line": 4,
            },
          },
          "name": "name",
          "value": "Custom story name",
        },
        Object {
          "loc": SourceLocation {
            "end": Position {
              "column": 3,
              "line": 10,
            },
            "start": Position {
              "column": 2,
              "line": 5,
            },
          },
          "name": "parameters",
          "value": Array [
            Object {
              "loc": SourceLocation {
                "end": Position {
                  "column": 28,
                  "line": 6,
                },
                "start": Position {
                  "column": 4,
                  "line": 6,
                },
              },
              "name": "component",
              "value": "ControlsTable",
            },
            Object {
              "loc": SourceLocation {
                "end": Position {
                  "column": 5,
                  "line": 9,
                },
                "start": Position {
                  "column": 4,
                  "line": 7,
                },
              },
              "name": "addonControls",
              "value": Array [
                Object {
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 18,
                      "line": 8,
                    },
                    "start": Position {
                      "column": 6,
                      "line": 8,
                    },
                  },
                  "name": "smart",
                  "value": false,
                },
              ],
            },
          ],
        },
      ],
      "source": "() => {}",
    },
  },
}
`;

exports[`csf-parameters story propery name parameters and controls 1`] = `
Object {
  "stories": Object {
    "myStory": Object {
      "arguments": Array [],
      "location": Object {
        "end": Object {
          "column": 31,
          "line": 2,
        },
        "start": Object {
          "column": 23,
          "line": 2,
        },
      },
      "name": "Custom story name",
      "properties": Array [
        Object {
          "loc": SourceLocation {
            "end": Position {
              "column": 27,
              "line": 4,
            },
            "start": Position {
              "column": 2,
              "line": 4,
            },
          },
          "name": "name",
          "value": "Custom story name",
        },
        Object {
          "loc": SourceLocation {
            "end": Position {
              "column": 3,
              "line": 17,
            },
            "start": Position {
              "column": 2,
              "line": 5,
            },
          },
          "name": "parameters",
          "value": Array [
            Object {
              "loc": SourceLocation {
                "end": Position {
                  "column": 5,
                  "line": 13,
                },
                "start": Position {
                  "column": 4,
                  "line": 6,
                },
              },
              "name": "controls",
              "value": Array [
                Object {
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 7,
                      "line": 12,
                    },
                    "start": Position {
                      "column": 6,
                      "line": 7,
                    },
                  },
                  "name": "name",
                  "value": Array [
                    Object {
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 31,
                          "line": 8,
                        },
                        "start": Position {
                          "column": 8,
                          "line": 8,
                        },
                      },
                      "name": "type",
                      "value": "ControlTypes.TEXT",
                    },
                    Object {
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 21,
                          "line": 9,
                        },
                        "start": Position {
                          "column": 8,
                          "line": 9,
                        },
                      },
                      "name": "label",
                      "value": "Name",
                    },
                    Object {
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 21,
                          "line": 10,
                        },
                        "start": Position {
                          "column": 8,
                          "line": 10,
                        },
                      },
                      "name": "value",
                      "value": "Mark",
                    },
                    Object {
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 19,
                          "line": 11,
                        },
                        "start": Position {
                          "column": 8,
                          "line": 11,
                        },
                      },
                      "name": "order",
                      "value": 9999,
                    },
                  ],
                },
              ],
            },
            Object {
              "loc": SourceLocation {
                "end": Position {
                  "column": 5,
                  "line": 16,
                },
                "start": Position {
                  "column": 4,
                  "line": 14,
                },
              },
              "name": "addonControls",
              "value": Array [
                Object {
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 18,
                      "line": 15,
                    },
                    "start": Position {
                      "column": 6,
                      "line": 15,
                    },
                  },
                  "name": "smart",
                  "value": false,
                },
              ],
            },
          ],
        },
      ],
      "source": "() => {}",
    },
  },
}
`;
